Joinpoint matching problems related to array types

kriegaex opened the pull request:
Fixes https://github.com/eclipse/org.aspectj/issues/24.

For now, this PR is just a draft, because it only contains a set of 4 regression tests, only one of which passes. The PR is meant to produce CI build results, making it easier for @aclement to collaborate.  Andy, maybe you know how to quickly fix the problems at hand if you can just check out the feature branch, run the tests and focus on the bugfix. I did the initial leg work for you already.

kriegaex committed:
Add regression tests for GitHub bug #24

Relates to https://github.com/eclipse/org.aspectj/issues/24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Expand regression tests for GitHub bug #24

More test cases were added for
  - multi-dimensional arrays,
  - primitive type arrays.

Relates to https://github.com/eclipse/org.aspectj/issues/24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Fix WildTypePattern.isArray

The method falsely determined that a one-dimensional array was not an
array due to a one-off bug.

Relates to https://github.com/eclipse/org.aspectj/issues/24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Simplify if-else in WildTypePattern.matchesExactlyByName

A simple boolean condition is enough.

Loosely relates to https://github.com/eclipse/org.aspectj/issues/24, but
actually it is just drive-by cosmetics.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Handle one- and multi-dimensional array return types correctly

Fixes https://github.com/eclipse/org.aspectj/issues/24, both the array
return type matching as such as well as matching dimensionality patterns
correctly. E.g., 'Foo*[]' is not the same as 'Foo*[][]'. This also works
correctly in combination with asterisks, even for primitive types, i.e.
'in*[][]' correctly matches a 2-dimensional array of 'int'.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex commented on org.aspectj.matcher/src/main/java/org/aspectj/weaver/UnresolvedType.java:
```diff
@@ -161,6 +161,10 @@ public final boolean isArray() {
 	}
 
+	public final int getDimensions() {
+		return signature.replaceAll("^(\\[*).*", "$1").length();
```
@aclement, do you think it would be better to eagerly (constructor) or lazily (first `getDimensions()` call) calculate this value once, assign it to a member and then just fetch it when needed, instead of doing the regex search & replace multiple times? It might not affect performance much, so I am not sure if you would rather spend CPU cycles or memory for an `int` instance field.

aclement replied:
My immediate reaction is that this would be overridden by resolved types and implemented more intelligently there, so it'll only be impacting folks asking getDimensions on unresolved types. It would increase the size of all unresolved types (and there could be quite a few of those at certains stages of weaving). Like you I'm skeptical of fixing if it isn't a known problem.

kriegaex commented:
Hm, after my changes now `*` does not match array types anymore, because e.g. `*[]` matches 1-dimensional and `*[][]` 2-dimensional arrays. That is a regression which I did not consider while fixing the array type bugs. I guess, my PR needs a little rework. Sorry for not running the unit tests locally, only the new integration tests.

kriegaex committed:
Improve array matching for all TypePattern subclasses

Relates to #24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
WildTypePattern: fix hashCode and toString methods

Especially 'hashCode' did not correspond to 'equals', disregarding
several fields, array dimension information being only one of them. This
led to parts of pointcuts being ignored, because they were regarded as
duplicates. Example:

execution(Foo* *(..)) && !execution(Foo*[] *(..))

Here, the negated pattern was falsely regarded as equal to the first
pattern, leading to an "A && !A" situation, i.e. no match at all.

Furthermore, 'toString' did not print array strings, i.e. instead of
"Foo*[][]" something like "Foo*" was printed. This false information was
also present in annotations generated by the weaver.

FuzzilyMatchingAspect was adjusted to actually match exactly once, as
expected, for the "Foo*" return types, i.e. exclusions for the array
return types have been added.

Relates to #24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Fix PointcutRewriterTest, add LogicalPointcutStructure test helper class

After WildTypePattern.hashCode was fixed in the previous commit,
PointcutRewriterTest started failing, because in many places it was
falsely relying on a specific order of hash codes, which cannot be
guaranteed, especially since more instance fields are part of the hash
code now in accordance with 'equals'.

The new test helper class LogicalPointcutStructure is able to recognise
chained '&&' and '||' pointcuts of the same logical nesting level,
un-nesting them from the actual pointcut structure and making them
comparable, disregarding their order. I.e., something like
  ((A && B) && C) && D
is actually recognised to logically be
  A && B && C && D
and equivalent to e.g. either of
  D && B && A && C
  A && B && D && C
  C && A && D && B

This helps to compare rewritten pointcuts, as long as their logical
structure has not been altered.

Relates to #24.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex commented:
The previous problem and a few others I detected along the way have been fixed.

While testing array type wildcard expressions, I noticed that **we are missing syntax for explicitly matching non-array types**. While for exact types, `FooBar` does not match `FooBar[]` or `FooBar[][]` types, in the wildcard case `Foo*` matches arrays, too. In order to exclude arrays, we have to do something like `execution(Foo* *(..)) && !execution(Foo*[] *(..)) && !execution(Foo*[][] *(..))`, because AspectJ is lacking a syntax such as `Foo*![]` or `Foo*[0]`. In order to add that, we would have to extend the grammar and the parser. Class `WildTypePattern` would also need a member to store the array exclusion or add special semantics to something like `getDimension()` returning -1 or so. For now, I chose not to do that, but it is something to keep in mind. WDYT, @aclement?

kriegaex committed:
Remove old, erroneously committed binaries from tests/bugs193/543657

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Fix failing tests after previous commit

After the changes, the weaver's behaviour is now slightly different, but
not actually buggy. Actually, in one case there is now a weaver warning
for a non-matching pointcut which should have been there before, but was
not. I.e., things have improved and the tests are even a bit better now.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>

kriegaex committed:
Comment on newly pointcut parsing found problem in test class

A pointcut like

  execution(*..Collection<?>[] *(..))

leads to an AJ core dump, which is something I noticed while fixing a
test in the previous commit. I am going to create a new issue for it and
link to this commit later.

Signed-off-by: Alexander Kriegisch <Alexander@Kriegisch.name>
